domain/src/DomainAccessControlHandler.php

```diff
@@ -18,25 +18,61 @@ class DomainAccessControlHandler extends EntityAccessControlHandler {
    * {@inheritdoc}
    */
   public function checkAccess(EntityInterface $entity, $operation, AccountInterface $account = NULL) {
+    $account = $this->prepareUser($account);
     // Check the global permission.
     if ($account->hasPermission('administer domains')) {
       return AccessResult::allowed();
     }
-    if ($operation == 'view' && ($entity->status() || $account->hasPermission('access inactive domains'))) {
+    // @TODO: This may not be relevant.
+    if ($operation == 'create' && $account->hasPermission('create domains')) {
       return AccessResult::allowed();
     }
-    if ($operation == 'create' && $account->hasPermission('create domains')) {
+    // For view, we allow unless the domain is inactive.
+    $is_admin = $this->isDomainAdmin($entity, $account);
+    if ($operation == 'view' && ($entity->status() || $account->hasPermission('access inactive domains'))) {
       return AccessResult::allowed();
     }
-    // @TODO: assign users to domains and check.
-    if ($operation == 'update' && $account->hasPermission('edit assigned domains')) {
+    // For other operations, check that the user is a domain admin.
+    if ($operation == 'update' && $account->hasPermission('edit assigned domains') && $is_admin) {
       return AccessResult::allowed();
     }
-    // @TODO: assign users to domains and check.
-    if ($operation == 'delete' && $account->hasPermission('delete assigned domains')) {
+    if ($operation == 'delete' && $account->hasPermission('delete assigned domains') && $is_admin) {
       return AccessResult::allowed();
     }
     return AccessResult::forbidden();
   }
 
+  /**
+   * Checks if a user can administer a specific domain.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity to retrieve field data from.
+   * @param \Drupal\Core\Session\AccountInterface
+   *   The user account.
+   *
+   * @return boolean
+   */
+  public function isDomainAdmin(EntityInterface $entity, AccountInterface $account) {
+    $user = \Drupal::entityTypeManager()->getStorage('user')->load($account->id());
```

Comment:
Looks like we want the pattern from /core/modules/node/src/NodeAccessControlHandler.php, which uses createInstance().